readthedocs/config/config.py

```diff
@@ -411,26 +481,103 @@ def validate_conf_file(self):
-
-        return True
+        return conf_file
 
     def validate_formats(self):
         """Validates that formats contains only valid formats."""
         formats = self.raw_config.get('formats')
         if formats is None:
-            return None
+            return self.defaults.get('formats', [])
         if formats == ['none']:
-            self['formats'] = []
-            return True
+            return []
 
         with self.catch_validation_error('format'):
             validate_list(formats)
             for format_ in formats:
                 validate_choice(format_, self.get_valid_formats())
-        self['formats'] = formats
 
-        return True
+        return formats
+
+    @property
+    def name(self):
+        """The project name."""
+        return self._config['name']
+
+    @property
+    def base(self):
+        """The base directory."""
+        return self._config['base']
+
+    @property
+    def output_base(self):
+        """The output base"""
+        return self._config['output_base']
+
+    @property
+    def type(self):
+        """The documentation type."""
+        return self._config['type']
+
+    @property
+    def formats(self):
+        """The documentation formats to be built."""
+        return self._config['formats']
+
+    @property
+    def python(self):
+        """Python related configuration."""
+        return self._config.get('python', {})
+
+    @property
+    def python_version(self):
+        """Python version."""
+        return self._config['python']['version']
+
+    @property
+    def pip_install(self):
+        """True if the project should be installed using pip."""
+        return self._config['python']['pip_install']
+
+    @property
+    def install_project(self):
+        """True if the project should be installed."""
+        if self.pip_install:
+            return True
+        return self._config['python']['setup_py_install']
+
+    @property
+    def extra_requirements(self):
+        """Extra requirements to be installed with pip."""
+        if self.pip_install:
+            return self._config['python']['extra_requirements']
+        return []
+
+    @property
+    def use_system_site_packages(self):
+        """True if the project should have access to the system packages."""
+        return self._config['python']['use_system_site_packages']
+
+    @property
+    def use_conda(self):
+        """True if the project use Conda."""
+        return self._config.get('conda') is not None
+
+    @property
+    def conda_file(self):
+        """The Conda environment file."""
+        if self.use_conda:
+            return self._config['conda'].get('file')
+        return None
+
+    @property
+    def requirements_file(self):
+        """The project requirements file."""
+        return self._config['requirements_file']
+
+    @property
+    def build_image(self):
+        """The docker image used by the builders."""
+        return self._config['build']['image']
```

Comment:
A couple things about the config attribute and usage pattern, which we can probably push off an resolve with a future PR:

* There is a lot of repetition here, for code that we're just using to convert a dict -> object with attributes
* We have multiple patterns for option naming: `python_version` vs `install_project`
* This pattern flattens some of the options, from `['python']['version']` to `python_version`, loss of nesting makes some of the options less obvious.

A couple of options here:

* Treat the config as a dictionary
* Rename some of the attributes now, so that they retain more contextual information -- ie, rename some of the nested options to `python_*`
* Use a namedtuple or a package like `bunch` to create nested attribute accessor instances, instead of repeating the logic above for each accessor

I think we can make a new issue here and address this later though.

Reply:
Yeah, I was thinking the same, maybe a namedtuple is enough, right now I just copy the names from the ConfigWrapper object.

Reply:
I opened #4346